fuzzbench/test_e2e/test_e2e_run.py

```diff
@@ -31,17 +41,29 @@ def redis_connection():
 # pylint: disable=no-self-use
 @pytest.mark.skipif('E2E_INTEGRATION_TEST' not in os.environ,
                     reason='Not running end-to-end test.')
-@pytest.mark.usefixtures('redis_connection')
+@pytest.mark.usefixtures('redis_connection', 'config')
 class TestEndToEndRunResults:
     """Checks the result of a test experiment run."""
 
-    def test_jobs_dependency(self):  # pylint: disable=redefined-outer-name
+    def test_jobs_dependency(self, config, redis_connection):  # pylint: disable=redefined-outer-name
         """Tests that jobs dependency preserves during working."""
-        assert True
+        all_images = docker_images.get_images_to_build(config['fuzzers'],
+                                                       config['benchmarks'])
+        jobs = {
+            name: rq.job.Job.fetch(name, connection=redis_connection)
+            for name in all_images
+        }
+        for name, image in all_images.items():
+            if 'depends_on' in image:
+                for dep in image['depends_on']:
+                    assert jobs[dep].ended_at <= jobs[name].started_at
```

Comment:
Need to figure out the differences among:
- enqueued_at
- started_at
- ended_at

Reply:
We also need to try putting a job first and put the its dependency later.
To check whether:
1. it will be located at the deferred_job_registry and wait for its dependency.
2. it will directly go into the queue for running.

Reply:
> We also need to try putting a job first and put the its dependency later.
> To check whether:
> 
> 1. it will be located at the deferred_job_registry and wait for its dependency.
> 2. it will directly go into the queue for running.

2. it will directly go into the queue for running.

Reply:
- enqueued_at: when the job enqueued to the queue
- started_at: when the job is worked by a worker
- ended_at: when the job is finished

Here is the execution result for 
```
jobs_list.append(queue.enqueue(jobs.run_trial, name='first', job_id='first'))
jobs_list.append(queue.enqueue(jobs.run_trial, name='second', job_id='second', depends_on='first'))
```
where one job is just a sleeping 10 seconds:
```
run-experiment_1  | All done!
run-experiment_1  |   fuzzbench.jobs.run_trial : finished       (first)
run-experiment_1  |    enqueued_at:  2020-08-03 18:53:29.681580
run-experiment_1  |    started_at:  2020-08-03 18:53:31.139904
run-experiment_1  |    ended_at:  2020-08-03 18:53:41.152006
run-experiment_1  |   fuzzbench.jobs.run_trial : finished       (second)
run-experiment_1  |    enqueued_at:  2020-08-03 18:53:41.155340
run-experiment_1  |    started_at:  2020-08-03 18:53:41.167681
run-experiment_1  |    ended_at:  2020-08-03 18:53:51.172041

Reply:
> > We also need to try putting a job first and put the its dependency later.
> > To check whether:
> > 
> > 1. it will be located at the deferred_job_registry and wait for its dependency.
> > 2. it will directly go into the queue for running.
> 
> it will directly go into the queue for running.

Here is the execution result for
```
jobs_list.append(queue.enqueue(jobs.run_trial, name='second', job_id='second', depends_on='first'))
jobs_list.append(queue.enqueue(jobs.run_trial, name='first', job_id='first'))
```
where one job is just a sleeping 10 seconds:
```
run-experiment_1  | All done!
run-experiment_1  |   fuzzbench.jobs.run_trial : finished       (first)
run-experiment_1  |    enqueued_at:  2020-08-03 18:57:50.513644
run-experiment_1  |    started_at:  2020-08-03 18:57:52.592289
run-experiment_1  |    ended_at:  2020-08-03 18:58:02.601616
run-experiment_1  |   fuzzbench.jobs.run_trial : finished       (second)
run-experiment_1  |    enqueued_at:  2020-08-03 18:57:50.512880
run-experiment_1  |    started_at:  2020-08-03 18:57:52.484071
run-experiment_1  |    ended_at:  2020-08-03 18:58:02.515156

Reply:
The above execution can be done through this branch: https://github.com/zchcai/fuzzbench/tree/dep-test-branch

Reply:
Thanks. So it sounds like we need to do a topological sort before enqueuing jobs. Currently, they happen to be in order, but please add a TODO to enforce this.